File: tests/js/langchain/test_reference.ts
/**
 * Conformance test: Reference instrumentation for LangChain.
 */
import { trace } from "@opentelemetry/api";
import { flushAndShutdownOtel, setupOtel } from "../otel";

const nodeProcess = globalThis.process as NodeJS.Process & {
  env: Record<string, string | undefined>;
  exit(code?: number): never;
};

const MOCK_BASE_URL = nodeProcess.env.MOCK_LLM_URL! + "/v1";
const tracer = trace.getTracer("gen_ai.reference");

async function main() {
  console.log("=== Reference: LangChain JS Conformance Test ===");
  const otel = setupOtel();

  const { ChatOpenAI, OpenAIEmbeddings } = await import("@langchain/openai");

  const chatModel = "gpt-4o-mini";
  const llm = new ChatOpenAI({
    model: chatModel,
    apiKey: "mock-key",
    configuration: { baseURL: MOCK_BASE_URL },
  });

  // Scenario: chat
  console.log("  [chat] basic chat completion (reference)");
  await tracer.startActiveSpan("chat gpt-4o-mini", async (span) => {
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "openai");
    span.setAttribute("gen_ai.request.model", chatModel);
    const resp = await llm.invoke([["user", "Say hello."]]);
    const responseMetadata = resp.response_metadata as any;
    if ((resp as any).id) {
      span.setAttribute("gen_ai.response.id", (resp as any).id);
    }
    if (responseMetadata?.model_name) {
      span.setAttribute("gen_ai.response.model", responseMetadata.model_name);
    }
    if (responseMetadata?.finish_reason) {
      span.setAttribute("gen_ai.response.finish_reasons", [responseMetadata.finish_reason]);
    }
    if (resp.usage_metadata) {
      span.setAttribute("gen_ai.usage.input_tokens", resp.usage_metadata.input_tokens);
      span.setAttribute("gen_ai.usage.output_tokens", resp.usage_metadata.output_tokens);
    }
    console.log(`    -> ${resp.content.toString().slice(0, 60)}`);
    span.end();
  });

  // Scenario: streaming chat
  console.log("  [chat_streaming] streaming chat completion (reference)");
  await tracer.startActiveSpan("chat gpt-4o-mini", async (span) => {
    span.setAttribute("gen_ai.operation.name", "chat");
    span.setAttribute("gen_ai.provider.name", "openai");
    span.setAttribute("gen_ai.request.model", chatModel);
    const stream = await llm.stream([["user", "Tell me a joke."]]);
    let text = "";
    let usageMetadata: any = null;
    let responseMetadata: any = null;
    let responseId: string | null = null;
    for await (const chunk of stream) {
      text += chunk.content;
      if (chunk.usage_metadata) usageMetadata = chunk.usage_metadata;
      if ((chunk as any).id) responseId = (chunk as any).id;
      if (chunk.response_metadata && Object.keys(chunk.response_metadata).length > 0) {
        responseMetadata = chunk.response_metadata;
      }
    }
    if (responseId) {
      span.setAttribute("gen_ai.response.id", responseId);
    }
    if (responseMetadata?.model_name) {
      span.setAttribute("gen_ai.response.model", responseMetadata.model_name);
    }
    if (responseMetadata?.finish_reason) {
      span.setAttribute("gen_ai.response.finish_reasons", [responseMetadata.finish_reason]);
    }
    if (usageMetadata) {
      span.setAttribute("gen_ai.usage.input_tokens", usageMetadata.input_tokens);
      span.setAttribute("gen_ai.usage.output_tokens", usageMetadata.output_tokens);
    }
    console.log(`    -> ${text.slice(0, 60)}`);
    span.end();
  });

  // Scenario: embeddings
  console.log("  [embeddings] embedding generation (reference)");
  await tracer.startActiveSpan("embeddings text-embedding-3-small", async (span) => {
    const requestModel = "text-embedding-3-small";
    span.setAttribute("gen_ai.operation.name", "embeddings");
    span.setAttribute("gen_ai.provider.name", "openai");
    span.setAttribute("gen_ai.request.model", requestModel);
    const embedModel = new OpenAIEmbeddings({
      model: requestModel,
      apiKey: "mock-key",
      configuration: { baseURL: MOCK_BASE_URL },
    });
    const result = await embedModel.embedQuery("Hello, world!");
    console.log(`    -> embedding dim: ${result.length}`);
    span.end();
  });

  await flushAndShutdownOtel(otel);
}

main().catch((e) => { console.error(e); nodeProcess.exit(1); });

File: tests/python/langchain/test_reference.py
"""Conformance test: reference instrumentation for LangChain."""

import os

from opentelemetry import trace

from otel_setup import flush_and_shutdown, setup_otel

MOCK_BASE_URL = os.environ["MOCK_LLM_URL"] + "/v1"

_reference_tracer = trace.get_tracer("gen_ai.reference")


def _usage_value(usage, key):
    if not usage:
        return None
    if isinstance(usage, dict):
        return usage.get(key)
    return getattr(usage, key, None)


def run_chat_reference(llm, request_model):
    """Scenario: basic chat completion with reference instrumentation."""
    print("  [chat] basic chat completion (reference)")
    with _reference_tracer.start_as_current_span("chat gpt-4o-mini") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "openai")
        span.set_attribute("gen_ai.request.model", request_model)
        resp = llm.invoke("Say hello.")
        meta = getattr(resp, "response_metadata", {})
        if meta.get("model_name"):
            span.set_attribute("gen_ai.response.model", meta["model_name"])
        if getattr(resp, "id", None):
            span.set_attribute("gen_ai.response.id", resp.id)
        if meta.get("finish_reason"):
            span.set_attribute("gen_ai.response.finish_reasons", [meta["finish_reason"]])
        usage = getattr(resp, "usage_metadata", None)
        input_tokens = _usage_value(usage, "input_tokens")
        output_tokens = _usage_value(usage, "output_tokens")
        if input_tokens is not None:
            span.set_attribute("gen_ai.usage.input_tokens", input_tokens)
        if output_tokens is not None:
            span.set_attribute("gen_ai.usage.output_tokens", output_tokens)
        print(f"    -> {resp.content[:60]}")


def run_chat_streaming_reference(llm, request_model):
    """Scenario: streaming chat completion with reference instrumentation."""
    print("  [chat_streaming] streaming chat completion (reference)")
    with _reference_tracer.start_as_current_span("chat gpt-4o-mini") as span:
        span.set_attribute("gen_ai.operation.name", "chat")
        span.set_attribute("gen_ai.provider.name", "openai")
        span.set_attribute("gen_ai.request.model", request_model)
        text = ""
        full = None
        for chunk in llm.stream("Tell me a joke."):
            text += chunk.content
            full = chunk if full is None else full + chunk
        if full:
            meta = getattr(full, "response_metadata", {})
            if meta.get("model_name"):
                span.set_attribute("gen_ai.response.model", meta["model_name"])
            if getattr(full, "id", None):
                span.set_attribute("gen_ai.response.id", full.id)
            if meta.get("finish_reason"):
                span.set_attribute("gen_ai.response.finish_reasons", [meta["finish_reason"]])
            usage = getattr(full, "usage_metadata", None)
            input_tokens = _usage_value(usage, "input_tokens")
            output_tokens = _usage_value(usage, "output_tokens")
            if input_tokens is not None:
                span.set_attribute("gen_ai.usage.input_tokens", input_tokens)
            if output_tokens is not None:
                span.set_attribute("gen_ai.usage.output_tokens", output_tokens)
        print(f"    -> {text[:60]}")


def run_embeddings_reference():
    """Scenario: embedding generation with reference instrumentation."""
    print("  [embeddings] embedding generation (reference)")
    from langchain_openai import OpenAIEmbeddings

    request_model = "text-embedding-3-small"
    embeddings = OpenAIEmbeddings(
        model=request_model,
        openai_api_base=MOCK_BASE_URL,
        openai_api_key="mock-key",
    )
    with _reference_tracer.start_as_current_span("embeddings text-embedding-3-small") as span:
        span.set_attribute("gen_ai.operation.name", "embeddings")
        span.set_attribute("gen_ai.provider.name", "openai")
        span.set_attribute("gen_ai.request.model", request_model)
        result = embeddings.embed_query("Hello, world!")
        print(f"    -> embedding dim: {len(result)}")


def main():
    print("=== Reference: LangChain Conformance Test ===")

    tp, lp, mp = setup_otel()

    from langchain_openai import ChatOpenAI

    request_model = "gpt-4o-mini"
    llm = ChatOpenAI(
        model=request_model,
        base_url=MOCK_BASE_URL,
        api_key="mock-key",
    )

    run_chat_reference(llm, request_model)
    run_chat_streaming_reference(llm, request_model)
    run_embeddings_reference()

    flush_and_shutdown(tp, lp, mp)


if __name__ == "__main__":
    main()
